smarts/sstudio/genhistories.py

```diff
@@ -277,34 +290,95 @@ def column_val_in_row(self, row, col_name: str):
+                                + np.sign(angular_velocity)
+                                * self._max_angular_velocity
+                                * 0.1
+                            )
+                    self._prev_heading = new_heading
+                    return new_heading
             # Note: pedestrian track files won't have this
-            return float(row.get("psi_rad", 0.0)) - math.pi / 2
+            self._prev_heading = float(row.get("psi_rad", 0.0)) - math.pi / 2
+            return self._prev_heading
         # XXX: should probably check for and handle x_offset_px here too like in NGSIM
         return None
 
 
 class NGSIM(_TrajectoryDataset):
-    def __init__(self, dataset_spec, output):
+    def __init__(self, dataset_spec: Dict[str, Any], output: str):
         super().__init__(dataset_spec, output)
-        self._prev_heading = -math.pi / 2
-
-    def _cal_heading(self, window):
-        c = window[1, :2]
-        n = window[2, :2]
-        if any(np.isnan(c)) or any(np.isnan(n)):
-            return self._prev_heading
-        s = np.linalg.norm(n - c)
-        if s == 0.0:
-            return self._prev_heading
-        vhat = (n - c) / s
-        r = math.atan2(vhat[1], vhat[0])
-        self._prev_heading = (r - math.pi / 2) % (2 * math.pi)
-        return self._prev_heading
+        self._prev_heading = 3 * math.pi / 2
+        self._max_angular_velocity = dataset_spec.get("max_angular_velocity", None)
+        self._heading_window = dataset_spec.get("heading_inference_window", 2)
+        # .22 corresponds to roughly 5mph.
+        self._heading_min_speed = dataset_spec.get("heading_inference_min_speed", 0.22)
+
+    def check_dataset_spec(self, dataset_spec: Dict[str, Any]):
+        super().check_dataset_spec(dataset_spec)
+        hiw = dataset_spec.get("heading_inference_window", 2)
+        if hiw < 2 or hiw > 11:
+            raise ValueError("heading_inference_window must be between 2 and 11")
+
+    def _cal_heading(self, window) -> float:
+        window = window[0]
+        new_heading = 0
+        prev_heading = None
+        den = 0
+        for w in range(self._heading_window - 1):
+            c = window[w, :2]
+            n = window[w + 1, :2]
+            if any(np.isnan(c)) or any(np.isnan(n)):
+                if prev_heading is not None:
+                    new_heading += prev_heading
+                    den += 1
+                continue
+            s = np.linalg.norm(n - c)
+            ispeed = window[w, 2]
+            if s == 0.0 or (
+                self._heading_min_speed is not None
+                and (s < self._heading_min_speed or ispeed < self._heading_min_speed)
+            ):
+                if prev_heading is not None:
+                    new_heading += prev_heading
+                    den += 1
+                continue
+            vhat = (n - c) / s
+            r = math.atan2(vhat[1], vhat[0])
+            inst_heading = (r - math.pi / 2) % (2 * math.pi)
```

Comment:
I believe our `math` should have this conversion method but it does not. (I believe it is being added in the OpenDRIVE pr)

Reply:
Yeah, actually I think I can just use `vec_to_radians()`.  It probably should have been that all along!